src/main/java/io/jenkins/plugins/ml/Server.java

```diff
@@ -0,0 +1,99 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package io.jenkins.plugins.ml;
+
+import hudson.Extension;
+import hudson.model.AbstractDescribableImpl;
+import hudson.model.Descriptor;
+import hudson.model.Job;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.annotation.Nonnull;
+
+public class Server extends AbstractDescribableImpl<Server> {
+    private String serverName;
+    private String serverAddress;
+    private long launchTimeout;
+    private long maxResults;
+
+    @DataBoundConstructor
+    public Server(String serverName, String serverAddress, long launchTimeout,long maxResults) {
+        this.serverName = serverName;
+        this.serverAddress = serverAddress;
+        this.launchTimeout = launchTimeout;
+        this.maxResults = maxResults;
+    }
+
+    String getServerName() {
+        return serverName;
+    }
+
+    public String getServerAddress() {
+        return serverAddress;
+    }
+
+    public long getLaunchTimeout() {
+        return launchTimeout;
+    }
+
+    public long getMaxResults() {
+        return maxResults;
+    }
+
+    @Override
+    public Descriptor<Server> getDescriptor() {
+        return null;
+    }
+
+    @Extension
+    public static class DescriptorImpl extends Descriptor<Server> {
+        @Nonnull
+        @Override
+        public String getDisplayName() {
+            return "Server";
+        }
+    }
+
+    /**
+     * Gets the {@link Server} associated with the given job.
+     *
+     * @return local host
+     *         if no such was found.
+     */
+    public static Server get(Job<?, ?> p) {
+        if(p != null) {
+            ServerJobProperty ijp = p.getProperty(ServerJobProperty.class);
+            if (ijp != null) {
+                // Looks in global configuration for the server configured
+                Server site = ijp.getSite();
+                if (site != null) {
+                    return site;
+                }
+            }
+        }
+        return new Server("localhost","127.0.0.1",1000,2);
```

Comment:
Maybe we should raise a runtime exception here instead?

It could lead to requests from the Jenkins server to 127.0.0.1:1000, which has no guarantee of actually having a running IPython server. In that case, I think it makes sense to say "Here's an error, because the program has no idea what to do"

Reply:
@kinow There is no API available to set the port in zeppelin-python API. It automatically searches only for the available port to connect the python instance.
```
String gatewayHost = PythonUtils.getLocalIP(this.properties);
int gatewayPort = RemoteInterpreterUtils.findRandomAvailablePortOnAllLocalInterfaces();
```
Here, even we configured as localhost. API does not use a used port. If we do not specify a default here, which his the best place to do the default? (In builder ? or somewhere?)

Reply:
I think we need to step back and look why the port would be missing.

In this method, from what I can tell it would do so iff the job was null.

That's not supposed to happen.

But if that happens, you have no way of knowing what port we should use.

If we choose any pprt number here, it could be wrong.

For me, the best approach would be to thrown an exception, and warn the user we couldn't configure the ipython server.

But we can wait to see what other mentors think of this too, as I could be wrong 👍

Reply:
We can throw the exception here as you suggested. And I just want to let you know @kinow the `1000` I set as a default, is not a port number; it is timeout for launcing the server.

Reply:
Aaaahhh! Sorry @Loghijiaha , misread your previous comment. 

So the server address is set to `127.0.0.1`, without a port number. Gotcha. Still better an exception IMHO.

Will try to read with more calm @Loghijiaha :-)